### ISSUE TITLE
Unusual behaviour in State type

### ISSUE THREAD

**Issue by katharine-murphy** [fetchai/ledger#1122]
Please find test code below. Come and have a chat if you need an in person demonstration - which may be more useful.

You will see that example 1, 2, and 3 work well and as expected.

Example 4 is picking up a value set outside the function, I assume with the key reference as that is the only link in the new `State`. We expect Example 4 to return 100, and instead it returns 42.

The offending line is : 
```// this now resets the value related to "balance" and is printed  at line 26
  var anotherState = State<Int32>("balance", 42);
```

(You can run Example 5 by uncommenting the code - example 5 demonstrates the effectiveness of the `set()` function).

Example 6 is an attempt to mimic the behaviour of Example 4 error coming from the `query()` function by creating a whole bunch of new `State` types in another function, but these have no effect at all on the original `State`.

```
function main()

  var myState = State<Int32>("balance", 0);
  // 1. print empty state
  print("1: ");
  printLn(toString(myState.get()));
  // PRINTS 0

  // 2. change state inside main
  myState.set(33);
  print("2: ");
  printLn(toString(myState.get()));
  // PRINTS 33

  change_state(myState);
  // 3. print state after change_state call
  print("3: ");
  printLn(toString(myState.get()));
  // PRINTS 100

  // ALL GOOD UP TO HERE....

  var result = query();
  // 4. print state after query call 
  print("4: ");
  printLn(toString(result));
  // PRINTS 42

  // 5. alter comments in query method to see that state value is not alterable even though
  // accessible from another function
  // print("5: ");
  // printLn(toString(result));
  // printLn(toString(myState.get()));
  // PRINTS 55

  // 6. Create new States with same key
  print("6: ");
  same_key();
  printLn(toString(myState.get()));

endfunction



function change_state(state : State<Int32>)

  // you can pass around a State parameter and get access to the original state
  var newState = state;
  newState.set(100);

  // this now resets the value related to "balance" and is printed  at line 26
  var anotherState = State<Int32>("balance", 42);

endfunction



function query() : Int32

  // this gives a default value, but it is not recognised 
  var myState = State<Int32>("balance", 0);

  // 4. and this returns 42!
  return myState.get();

  // 5. 
  // If you set a new variable into the state, it will return the new variable 
  // myState.set(55);
  // return myState.get();

endfunction



function same_key()

  // these have no effect on original state
  var myState1 = State<Int32>("balance", 32);
  var myState = State<Int32>("balance", 33);
  var myState2 = State<Int32>("balance", 34);
  var myState3 = State<Int32>("balance", 35);
  var myState4 = State<Int32>("balance", 36);
  var myState5 = State<Int32>("balance", 37);

endfunction
```

**Comment by katharine-murphy** [comment]
I think this is fixed in `v0.4.0-alpha3-2-g827a6397` and I was working on old ledger`v0.4.0-alpha3`.

**Comment by katharine-murphy** [comment]
Yes, looks OK now. Still need to clarify what the string key does. Going to close this issue.
